### PR TITLE
Spacecraft Modules Stack

### PR DESCRIPTION
**Pull request by Pedro-Roque**
PR directed to adding all functionality for spacecrafts using PX4.

To-does:
- feed ground-truth to mocap odometry and test posctl in SITL;
- ~Actuator testing does not output anything on the actuator_motors topic~ Fixed. A future PR to https://github.com/mavlink/qgroundcontrol/blob/5967a9e888891f26f62e8c33b95b18255928f5c0/src/Vehicle/Actuators/ActuatorTesting.cc#L48 could be a nice addition to all testing All Thrusters.
- QGC complains about a missing parameter FW_ARSP_MODE: needs to be fixed via QGC proper airframe definition.
- ~In SITL, setting Manual mode triggers comm loss;~ Not present anymore. Arming via pxh> works, also works in manual mode. Rate setpoints are set and thrust/torque setpoints are obtained.

@Jaeyoung-Lim ideally we should work on this branch or towards merging things on this branch before merging in main.

### CODE REVIEW

**Comment by Pedro-Roque** [comment]
Build currently failing with
```bash
Exception while parsing /home/roque/ros2_astrobee_ws/src/submodules/px4/src/modules/control_allocator/module.yaml:
Traceback (most recent call last):
  File "/home/roque/ros2_astrobee_ws/src/submodules/px4/Tools/module_config/generate_actuators_metadata.py", line 484, in <module>
    raise e
  File "/home/roque/ros2_astrobee_ws/src/submodules/px4/Tools/module_config/generate_actuators_metadata.py", line 476, in <module>
    parsed_mixers = get_mixers(yaml_config, output_functions, verbose)
  File "/home/roque/ros2_astrobee_ws/src/submodules/px4/Tools/module_config/generate_actuators_metadata.py", line 307, in get_mixers
    output_function = output_functions['common'][function_name]
KeyError: 'Thruster'
```

Will fix it later.

**Comment by Pedro-Roque** [comment]
Thrusters changes reflected all the way in QGC, but arming is still denied. Will attempt to fix that.
Screenshot from 2024-04-22 19-01-27

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessSpacecraft.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	const bool thrusters_added_successfully = _sc_thrusters.addActuators(configuration);
+
+	return thrusters_added_successfully;
+}
+
+void ActuatorEffectivenessSpacecraft::updateSetpoint(const matrix::Vector<float, NUM_AXES> &control_sp,
+		int matrix_index, ActuatorVector &actuator_sp, const matrix::Vector<float, NUM_ACTUATORS> &actuator_min,
+		const matrix::Vector<float, NUM_ACTUATORS> &actuator_max)
+{
+	// TODO(@E-Krantz): Here is where we can add the nonlinearity of multiple thrusters open
```
Comment (Pedro-Roque) [inline src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessSpacecraft.cpp:62]: @E-Krantz Note where we can then adapt the setpoints based on the thrust to unit mapping and add nonlinearity based on how many thrusters are open and by how much

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessSpacecraft.hpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+	void getDesiredAllocationMethod(AllocationMethod allocation_method_out[MAX_NUM_MATRICES]) const override
+	{
+		// TODO(@Jaeyoung-Lim): to be updated
+		allocation_method_out[0] = AllocationMethod::SEQUENTIAL_DESATURATION;
+	}
+
+	void getNormalizeRPY(bool normalize[MAX_NUM_MATRICES]) const override
+	{
+		// TODO(@Jaeyoung-Lim): to be updated
```
Comment (Pedro-Roque) [inline src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessSpacecraft.hpp:55]: @Jaeyoung-Lim

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessThrusters.hpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+	ActuatorEffectivenessThrusters(ModuleParams *parent, AxisConfiguration axis_config = AxisConfiguration::Configurable,
+				    bool tilt_support = false);
+	virtual ~ActuatorEffectivenessThrusters() = default;
+
+	bool getEffectivenessMatrix(Configuration &configuration, EffectivenessUpdateReason external_update) override;
+
+	void getDesiredAllocationMethod(AllocationMethod allocation_method_out[MAX_NUM_MATRICES]) const override
+	{
+          allocation_method_out[0] = AllocationMethod::SEQUENTIAL_DESATURATION;  // TODO(@Jaeyoung-Lim): needs to be updated based on metric mixer
```
Comment (Pedro-Roque) [inline src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessThrusters.hpp:82]: @Jaeyoung-Lim

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessThrusters.hpp [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
+	void getNormalizeRPY(bool normalize[MAX_NUM_MATRICES]) const override
+	{
+		// TODO(@Jaeyoung-Lim): needs to be updated based on metric mixer
```
Comment (Pedro-Roque) [inline src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessThrusters.hpp:87]: @Jaeyoung-Lim

**Comment by Pedro-Roque** [comment]
@Jaeyoung-Lim Sorry for the spam - I tagged you where in the code we would take your input. Right now the mixer seems to be working, but I'll be testing it with a few position setpoints.

@E-Krantz I tagged you where I believe you would add the code that modifies the setpoints based on how many thrusters are requested to be open (probably you will have some adjustment to ask for more thrust when more thrusters are used, to have effectively the nominal thruster behavior)

I'll see if I get this to run with attitude and position control :smile:

**Comment by Pedro-Roque** [comment]
Some problems to note:
1. Actuator testing does not output anything on the actuator_motors topic
2. QGC complains about a missing parameter FW_ARSP_MODE (however this is note used anywhere)
3. In SITL, setting Manual mode triggers comm loss;

**Comment by Pedro-Roque** [comment]
Control pipeline from att control through the rate, allocation and mixer seems alive. Needs testing.

Next step will be to add SC Position Control.

**Comment by Pedro-Roque** [comment]
@Jaeyoung-Lim I merged everything into this branch - you can use it to develop the Gazebo plugin. Branch out of this one and then merge back into it.

**Comment by Pedro-Roque** [comment]
To test: feed ground-truth to mocap odometry and test posctl in SITL.

**Comment by Pedro-Roque** [comment]
Big Pile of Changes from upstream sync...

**Comment by Pedro-Roque** [comment]
@Jaeyoung-Lim a thing that is annoying now is that without without rate control running, the simulator stalls, so we can't test Manual mode; We should really consider moving the lockstep elsewhere...

**Comment by Jaeyoung-Lim** [comment]
@Pedro-Roque Where would you move it? Fundamentally, direct actuator control does not really work with lockstep, because the actuator commands do not correlate with sensor data.

We could also just turn lockstep off for testing actuator controls.

I am not really motivated to fix this " properly" for gazebo classic, as we fixed this in the new gazebo with a better architecture, and gazebo classic will not last more than a year or two...

**Pull request merged**